### PR TITLE
Refactor IO utils

### PR DESCRIPTION
**Pull request by MiSikora**
## :page_facing_up: Context

Just some cleanup of IO utils that was done along the way of #527. Moving methods out of `IOUtils` class to some extensions functions etc.

## :pencil: Changes

As mentioned. It is mostly moving internal parts that are about IO and HTTP to separate classes.

## :paperclip: Related PR

#527

## :no_entry_sign: Breaking

No.

## :hammer_and_wrench: How to test

Nothing changed in that regard. Tests were adjusted accordingly.

### CODE REVIEW

**Review by vbuberen — APPROVED** [review]

File: library/src/test/java/com/chuckerteam/chucker/internal/support/OkioUtilsTest.kt [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import okio.ByteString.Companion.encodeUtf8
+import org.junit.jupiter.api.Test
+import java.nio.charset.Charset
+
+internal class OkioUtilsTest {
+    @Test
+    fun isPlaintext_withEmptyBuffer_returnsTrue() {
+        val buffer = Buffer()
+
+        Truth.assertThat(buffer.isProbablyPlainText).isTrue()
```
Comment (vbuberen) [inline library/src/test/java/com/chuckerteam/chucker/internal/support/OkioUtilsTest.kt:14]: Nitpick: why don't we import this function for consistency with other tests?

Comment (MiSikora) [reply library/src/test/java/com/chuckerteam/chucker/internal/support/OkioUtilsTest.kt:14]: Yup, thanks! Saw it as well and forgot to change it 😅 .

File: library/src/main/java/com/chuckerteam/chucker/internal/support/OkioUtils.kt [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ */
+internal val Buffer.isProbablyPlainText
+    get() = try {
+        val prefix = Buffer()
+        val byteCount = if (size < MAX_PREFIX_SIZE) size else MAX_PREFIX_SIZE
+        copyTo(prefix, 0, byteCount)
+        sequence { while (!prefix.exhausted()) yield(prefix.readUtf8CodePoint()) }
+            .take(CODE_POINT_SIZE)
+            .any { codePoint -> Character.isISOControl(codePoint) && !Character.isWhitespace(codePoint) }
+            .not()
```
Comment (cortinico) [inline library/src/main/java/com/chuckerteam/chucker/internal/support/OkioUtils.kt:21]: I have to admit that the logic here is a bit hard to follow due to `!` and `.not`. Could we reword it to a single expression using `.all{}` or `.none{}`?

File: library/src/test/java/com/chuckerteam/chucker/internal/support/OkioUtilsTest.kt [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+import com.google.common.truth.Truth.assertThat
+import okio.Buffer
+import okio.ByteString.Companion.encodeUtf8
+import org.junit.jupiter.api.Test
+import java.nio.charset.Charset
+
+internal class OkioUtilsTest {
+    @Test
+    fun isPlaintext_withEmptyBuffer_returnsTrue() {
```
Comment (cortinico) [inline library/src/test/java/com/chuckerteam/chucker/internal/support/OkioUtilsTest.kt:11]:
```suggestion
    fun isProbablyPlainText_withEmptyBuffer_returnsTrue() {
```

This applies to all the other tests

File: library/src/test/java/com/chuckerteam/chucker/internal/support/OkHttpUtilsTest.kt [hunk 4]
```diff
@@ -66,4 +76,50 @@ internal class OkHttpUtilsTest {
...
+    fun headersHaveSupportedEncoding(headers: Headers, isSupported: Boolean) {
+        val result = headers.hasSupportedContentEncoding
+
+        assertThat(result).isEqualTo(isSupported)
+    }
+
+    companion object {
+        @JvmStatic
+        fun supportedEncodingSource(): Stream<Arguments> = Stream.of(
+            "" to true,
```
Comment (cortinico) [inline library/src/test/java/com/chuckerteam/chucker/internal/support/OkHttpUtilsTest.kt:117]: we lost `null to true` here. I supposed that's due to a Kotlin API right?

File: library/src/main/java/com/chuckerteam/chucker/internal/support/OkHttpUtils.kt [hunk 5]
```diff
@@ -60,3 +63,17 @@ private val Headers.containsGzip: Boolean
     get() {
         return this["Content-Encoding"].equals("gzip", ignoreCase = true)
     }
+
+private val supportedEncodings = listOf("identity", "gzip")
+
+internal val Headers.hasSupportedContentEncoding: Boolean
+    get() = get("Content-Encoding")
+        ?.takeIf { it.isNotEmpty() }
+        ?.let { it.toLowerCase(Locale.ROOT) in supportedEncodings }
```
Comment (cortinico) [inline library/src/main/java/com/chuckerteam/chucker/internal/support/OkHttpUtils.kt:72]: nit: In another part of the codebase we use `Locale.US`. I believe we should switch to `.ROOT` also the other but I'm unsure.

Comment (MiSikora) [reply library/src/main/java/com/chuckerteam/chucker/internal/support/OkHttpUtils.kt:72]: The other code uses it for formatting. Leaving `US` there is fine.

https://github.com/ChuckerTeam/chucker/blob/b3761b7ceca25386624520fe49a5cbdb713b4a8f/library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt#L52

File: library/src/test/java/com/chuckerteam/chucker/internal/support/OkHttpUtilsTest.kt (hunk 4 above)
Comment (MiSikora) [reply library/src/test/java/com/chuckerteam/chucker/internal/support/OkHttpUtilsTest.kt:117]: Yup, brought it back.

File: library/src/main/java/com/chuckerteam/chucker/internal/support/OkioUtils.kt (hunk 2 above)
Comment (MiSikora) [reply library/src/main/java/com/chuckerteam/chucker/internal/support/OkioUtils.kt]: Simplified :).

**Review by cortinico — APPROVED** [review]
Great work 👌 Thanks for splitting the other PR 🙏

**Comment by MiSikora** [comment]
> Great work 👌 Thanks for splitting the other PR 🙏

It isn't finished yet. Will continue. ;)
